sycl/doc/extensions/proposed/sycl_ext_oneapi_annotated_ptr.asciidoc

```diff
@@ -0,0 +1,668 @@
+constant information `I` for a pointer. One possible method to do so is for the
+compiler to provide a special function for reading to and writing from the
+pointer and have this function convey `I` to the compiler at the call site where
+the read or write occurs.
+
+That is, users would convert code that looks like:
+```c++
+{
+  SomeType* a;
+  ...
+  *a = some_val;
+  ...
+  read_val = *a
+}
+```
+
+To code that looks like:
+```c++
+{
+  SomeType* a;
+  ...
+  __special_store_with_annotation(a, some_val,
+    I /*some compile time constant information*/);
+  ...
+  __special_load_with_annotation(a, &read_val,
+    I /*some compile time constant information*/);
+}
+```
+
+This mechanism does not meet requirement (1) listed above because users have to
+replace all their pointer read/write code with special function calls.
+
+Another mechanism could be that the compiler provides attributes which can be
+applied to the pointer declaration to convey some compile-time constant
+information. In this case users would change their code to:
+```cpp
+{
+  // 'I' is the compiler time constant information being annotated on 'a'
+  [[special_annotations::use_case_x_attribute(I)]] SomeType* a;
+  ...
+  *a = some_val;
+  ...
+  read_val = *a
+}
+```
+However, this mechanism fails to meet requirements (2) and (3) listed above
+since:
+
+  * Adding new attributes can add non-trivial attribute creation overheads for
+  compiler developers. With growing set of attributes, there is an associated
+  burden of maintaining them and of preparing diagnostic checks.
+  * Compilers are not required to preserve information propagated as attributes.
+  Hence this method is only suitable for hints, and not functional directives.
+  Further, from a given load/store call a compiler may not be able to
+  successfully trace back to the declaration to deduce which annotation was
+  applied to it.
+
+The `annotated_ptr` class described in this document is a class template that
+encapsulates a pointer. The template accepts a list of compile-time constant
+properties. The implementation of the class defined here should preserve the
+information provided as compile-time constant properties through all uses of the
+pointer unless noted otherwise.
+
+
+=== Representation of `annotated_ptr`
+
+`annotated_ptr` is a class template, parameterized by the type of the underlying
+allocation `T`, and a list of properties `PropertyListT`.
+
+[source,c++]
+----
+namespace sycl::ext::oneapi::experimental {
+template <typename T, typename PropertyListT = properties<>>
+class annotated_ptr {
+  ...
+----
+
+`PropertyListT` enables properties to be associated with an `annotated_ptr`.
+Properties may be specified for an `annotated_ptr` to provide semantic
+modification or optimization hint information.
+
+`T` can be any type except the following types or a structure containing one of
+the following types: sycl::accessor, sycl::stream, sycl::sampler, and
+sycl::half_type.
```

Comment:
I was not thinking that we needed this restriction here.  Do we?

Reply:
On second thought no. I was thinking from the perspective of sycl special class before. But pointer to these types are just that - pointers. So no decompositions are needed. I don't think we need these restrictions. Removed.

Made me wonder though: what does it mean to have a pointer to an accessor? Or to a stream? Are such arguments useful in any cases?

Reply:
It does seem weird to have a pointer to an accessor, but I can't think of any reason to disallow it.